woocommerce/woocommerce-ios
# [Manage Categories] Category selector screen pagination strategy

References: https://github.com/woocommerce/woocommerce-ios/issues/2000

# Why

In preparation for the Category Selector Screen view.

![Screen Shot 2020-04-11 at 5 47 50 PM](https://user-images.githubusercontent.com/562080/79056471-98a37880-7c1c-11ea-950a-b3fd745f5443.png)

This PR takes care of implementing a pagination strategy that makes sure that all product categories are fetched before displaying them to the user. The pagination is implemented in such way due to the limitation of the API on returning the product categories in our required order. (Parent, then children)

Additionally this PR adds ghost placeholders while synchronizing categories and displays an error notice(with a retry action) if the synchronize errors at some point.

# How
-  Modifies the `ProductCategoryAction.synchronizeCategories` to return the specific fetched product categories, to be able to know when we have finished fetching categories, which is when the fetched categories are empty.

- Adds a recursive method -`synchronizeAllCategories`- to fetch all categories before notifying the view controller.

- Adds a `syncCategoriesState` property in the view model, which the view controller can observe to get "synchronize categories" state updates and know when to display/remove the ghost placeholders or the error notice.

## Notes
While testing the pagination strategy I found out that the API only allow us to send a max number of `100` in the `pageSize` parameter. Sending a bigger number will make the network request to fail.


# Testing steps
- Log into a site
- Go to the products tab
- Tap on a product
- Tap on the categories field
- See that all categories are loaded in one go

# Demo

![DownloadAllCategories](https://user-images.githubusercontent.com/562080/79056893-becb1780-7c20-11ea-86cf-28e1645f0670.gif)
